File: algorithms/2_medium/L/01946_largest_number_after_mutating_substring/solution.cpp
constexpr size_t CHANGE_LEN = 10;

class Solution {
    static inline bool can_improve_by_change(const vector<int>& change_vec) {
        for (size_t i = 0; i < CHANGE_LEN; i += 1) {
            if (i < change_vec[i]) {
                return true;
            }
        }
        return false;
    }
public:
    static string maximumNumber(string num, const vector<int>& change_vec) {
        assert(change_vec.size() == CHANGE_LEN);

        if (!can_improve_by_change(change_vec)) {
            return num;
        }

        char change[CHANGE_LEN];
        for (size_t i = 0; i < CHANGE_LEN; i += 1) {
            const int x = change_vec[i];
            assert(x >= 0 && x <= 9);
            change[i] = static_cast<char>('0' + change_vec[i]);
        }

        bool has_changed = false;
        for (char& ch : num) {
            assert(ch >= '0' && ch <= '9');
            const size_t idx = static_cast<size_t>(ch - '0');
            if (ch < change[idx]) {
                ch = change[idx];
                has_changed = true;
            } else if (has_changed && ch != change[idx]) {
                break;
            }
        }

        return num;
    }
};

File: algorithms/2_medium/L/01946_largest_number_after_mutating_substring/solution.rs
impl Solution {
    const CHANGE_LEN: usize = 10;

    pub fn maximum_number(num: String, change_vec: Vec<i32>) -> String {
        assert!(change_vec.len() == Self::CHANGE_LEN);

        if (0..Self::CHANGE_LEN).into_iter().all(|x| (x as i32) >= change_vec[x]) {
            return num;
        }

        let mut change = [0_u8; Self::CHANGE_LEN];
        for i in 0..Self::CHANGE_LEN {
            let x = change_vec[i];
            assert!(x >= 0 && x <= 9);
            change[i] = b'0' + change_vec[i] as u8;
        }

        let mut bytes = num.into_bytes();
        let mut has_changed = false;
        for b in bytes.iter_mut() {
            assert!(*b >= b'0' && *b <= b'9');
            let idx = ((*b) - b'0') as usize;
            if *b < change[idx] {
                *b = change[idx];
                has_changed = true;
            } else if has_changed && *b != change[idx] {
                break;
            }
        }

        String::from_utf8(bytes).unwrap()
    }
}
